### PR TITLE
Sequence load 3 (missing test init)

### PR DESCRIPTION
**Pull request by hepcat72**
## Summary Change Description

I discovered that the new subdirectory in the tests directory wasn't getting any tests run, because I hadn't added `__init__.py`.  I had initially run those tests explicitly and they worked, but changes since then should have broken them, so when all the tests passed, I investigated and found the omission.  This PR represents a catch-up of the broken tests.

## Affected Issues/Pull Requests

- Partially addresses #824
- Merges into #869
- Next PR: #871

## Review Notes

See comments in-line.

## Checklist

This pull request will be merged once the following requirements are met.  The
author and/or reviewers should uncheck any unmet requirements:

- Review requirements
  - Minimum approvals: 1 
  - No changes requested
  - All blocking issues resolved by reviewers
  - Specific reviewers: @__add_username_here__
    
  - Review period: 2 days 
- Associated issue/pull request requirements:
  
  - [x] All requirements in affected issues marked "resolved" are satisfied
  - [x] All required pull requests are merged *(or none)*
- Basic requirements
  
  - [x] [All linters pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting)
  - [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All conflicts resolved
- Overhead requirements
  
  - [x] [New/changed method tests implemented/updated *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [ ] [Updated *Unreleased* section of `changelog.md` *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md)
  - [x] [Migrations created & committed *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)

### CODE REVIEW

File: DataRepo/tests/utils/test_loader.py [hunk 1]
```diff
@@ -478,13 +486,39 @@ def test_isnamedtuple(self):
...
         self.assertTrue(hasattr(tl, "all_headers"))
         self.assertTrue(hasattr(tl, "reqd_headers"))
         self.assertTrue(hasattr(tl, "FieldToHeader"))
         self.assertTrue(hasattr(tl, "unique_constraints"))
         self.assertTrue(hasattr(tl, "record_counts"))
+        self.assertTrue(hasattr(tl, "defaults_current_type"))
+        self.assertTrue(hasattr(tl, "sheet"))
+        self.assertTrue(hasattr(tl, "defaults_sheet"))
+        self.assertTrue(hasattr(tl, "reqd_values"))
+        self.assertTrue(hasattr(tl, "defaults_by_header"))
```
Comment (hepcat72) [inline DataRepo/tests/utils/test_loader.py:521]: C. Tests revealed some broken attribute values that I realized should have been caught here.

File: DataRepo/tests/utils/test_loader.py [hunk 2]
```diff
@@ -631,8 +667,6 @@ def test_tableheaders_to_dict_by_header_name(self):
     # Test that load_wrapper
     def test_load_wrapper_does_not_nest_AggregatedErrors(self):
         class TestNestedAesLoader(self.TestLoader):
-            # Must explicitly re-add the decorator when the derived class is used as a superclass
-            @TraceBaseLoader._loader
```
Comment (hepcat72) [inline DataRepo/tests/utils/test_loader.py:635]: C. These decorators should never have been here.  Double-derived classes still run the automatic application of the wrappers.  It just took me some time to realize that.

File: DataRepo/utils/loader.py [hunk 3]
```diff
@@ -236,7 +236,7 @@ def apply_loader_wrapper(self):
             Nothing
         """
         # Apply the _loader decorator to the load_data method in the derived class
-        decorated_derived_class_method = self._loader(getattr(self, "load_data"))
+        decorated_derived_class_method = self._loader()(getattr(self, "load_data"))
```
Comment (hepcat72) [inline DataRepo/utils/loader.py:239]: C. I changed the decorator to include the typical decorator method.  The code I found online had skipped it, and made the errors really confusing, especially with the doubly decorated methods I fixed above.

File: DataRepo/utils/loader.py [hunk 4]
```diff
@@ -704,6 +705,23 @@ def check_class_attributes(cls):
...
+                    if not issubclass(mdl, Model):
+                        mdlerrs.append(
+                            f"{type(mdl).__name__}: Not a subclass of a Django Model"
+                        )
+                if len(mdlerrs) > 0:
+                    nltt = "\n\t\t"
+                    typeerrs.append(
+                        "Models must all be valid models, but the following errors were encountered:\n"
+                        f"\t\t{nltt.join(mdlerrs)}"
+                    )
```
Comment (hepcat72) [inline DataRepo/utils/loader.py:724]: C. Moved from `initialize_metadata`.  It belonged here, as it's a class attribute, not a member attribute.

File: DataRepo/utils/loader.py [hunk 5]
```diff
@@ -1333,105 +1338,119 @@ def get_defaults_dict_by_header_name(self):
...
         Returns:
-            load_wrapper (function)
+            load_decorator (function)
         """
 
-        def load_wrapper(self, *args, **kwargs):
-            """Wraps the load_data() method in the derived class.
-
-            Checks the file data and handles atomic transactions, running modes, exceptions, and load stats.
+        def load_decorator(fn):
```
Comment (hepcat72) [inline DataRepo/utils/loader.py:1355]: C. The inserted decorator method I mentioned above.

File: DataRepo/utils/loader.py [hunk 6]
```diff
@@ -1333,105 +1338,119 @@ def get_defaults_dict_by_header_name(self):
...
 
             Raises:
-                TBD by derived class
+                Nothing
 
             Returns:
-                What's returned by the wrapped method
+                apply_wrapper (function)
             """
-            retval = None
```
Comment (hepcat72) [inline DataRepo/utils/loader.py:1365]: C. The rest of this method is mostly an indent change.

**Review by lparsons — APPROVED** [review]
LGTM.

**Comment by hepcat72** [comment]
You know, I was initially surprised that nothing had warned me about the missing `__init__.py` file.  Later, I realized that NOT throwing an error about something like that makes sense, but it made me wonder if there was a tool (e.g. a linter or analysis tool) that may have some sort of "strict" mode to catch things like that, like there is a `mypy` option to "check untyped functions" (which I construe as a form of "strict checking").  I was thinking that if something did exist to warn about init files, I could incorporate it into my own utility scripts.  I did a cursory search, but did not find anything.  Are you aware of anything like that @lparsons?

**Comment by lparsons** [comment]
> I was thinking that if something did exist to warn about init files, I could incorporate it into my own utility scripts. I did a cursory search, but did not find anything. Are you aware of anything like that @lparsons?

I do not know of an exiting toool, but I believe that [`__init__.py` files are not actually required in Python 3](https://stackoverflow.com/questions/37139786/is-init-py-not-required-for-packages-in-python-3-3) anymore. There is an issue that some tools (e.g. mypy) won't find those packages. I'm guessing a better long term "fix" would be to have the test finding logic not miss those tests, but I'm not sure of the issues around that either.

**Comment by hepcat72** [comment]
> I believe that [`__init__.py` files are not actually required in Python 3](https://stackoverflow.com/questions/37139786/is-init-py-not-required-for-packages-in-python-3-3) anymore. [...] I'm guessing a better long term "fix" would be to have the test finding logic not miss those tests, but I'm not sure of the issues around that either.

I have looked at the core code that finds the tests (that's what my `list_tests.py` script utilizes).  `django.test` has a lot of its own code, but it is still directly dependent on `pytest` for finding tests.  I don't know if perhaps there's a newer (django-compatible) version of that package, but if there is, maybe all we need is a dependency upgrade.

**Pull request merged**